edgedb/retry.py

```diff
@@ -0,0 +1,163 @@
+        self._done = False
+        self._backoff = default_backoff
+        self._max_iterations = DEFAULT_MAX_ITERATIONS
+
+    def __aiter__(self):
+        return self
+
+    async def __anext__(self):
+        if self._done:
+            raise StopAsyncIteration
+        assert self._iteration + 1 < self._max_iterations, \
+            f"Extra retry {self._iteration}/{self._max_iterations}"
+        if self._iteration > 0:
+            await asyncio.sleep(self._backoff(self._iteration))
+        self._done = True
+        iteration = AsyncIOIteration(self, self._owner, self._iteration)
+        self._iteration += 1
+        return iteration
+
+    def _retry(self, exc):
+        self._last_exception = exc
+        if self._iteration >= self._max_iterations:
+            return False
+        self._done = False
+        return True
+
+
+class Retry:
+
+    def __init__(self, owner):
+        self._owner = owner
+        self._iteration = 0
+        self._done = False
+        self._backoff = default_backoff
+        self._max_iterations = DEFAULT_MAX_ITERATIONS
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        if self._done:
+            raise StopIteration
+        assert self._iteration + 1 < self._max_iterations, \
+            f"Extra retry {self._iteration}/{self._max_iterations}"
+        if self._iteration > 0:
+            time.sleep(self._backoff(self._iteration))
+        self._done = True
+        iteration = Iteration(self, self._owner, self._iteration)
+        self._iteration += 1
+        return iteration
+
+    def _retry(self, exc):
+        self._last_exception = exc
+        if self._iteration >= self._max_iterations:
+            return False
+        self._done = False
+        return True
+
+
+class Iteration(_transaction.Transaction):
+    def __init__(self, retry, owner, iteration):
+        super().__init__(owner)
+        self.__retry = retry
+        self.__iteration = iteration
+
+    def start(self):
+        if not self._managed:
+            raise errors.InterfaceError(
+                "Only managed retriable transactions are supported. "
+                "Use `with transaction:`"
+            )
+        self._start(single_connect=self.__iteration != 0)
+
+    def __exit__(self, extype, ex, tb):
+        try:
+            super().__exit__(extype, ex, tb)
+        except errors.ClientConnectionClosedError as err:
+            if ex is None:
+                # TODO(tailhook) should we figure out if err is *caused by*
+                # CancelledError and propagate the latter instead or retrying?
+
+                # If we were going to commit, retry
+                if (
+                    err.has_tag(errors.SHOULD_RETRY) and
+                    self.__retry._retry(err)
+                ):
+                    return True
+                else:
+                    # if retries are exhausted we need to propagate this error
+                    raise err
```

Comment:
We need to make sure that we never mask CancelledError. It should be propagated as is.

Reply:
We don't mask anything here. We reraise same error and only catch EdgeDB errors anyway. Also see comment above: this code is now simplified similarly to the async case.